### ISSUE TITLE
Alternative training method

### ISSUE THREAD

**Issue by Grant-Tao** [leela-zero/leela-zero#2561]
A big difference between the training methods of MuZero and AlphaZero is:  Muzero uses K consequential  steps samples, and AlphaZero uses only one step samples. Muzero's paper has shown the advantage of this training approach. 

My suggestion is this:

Modify chunkparser.py, let parse() generate K batches data once, the samples of this K batches form one batch of consequential K positions of different games 

Modify tfprocess.py, using gradient accumulation method, updating network parameters with the averaged gradients of K batches.

In MuZero, K=5

I do not have the resource  for training, maybe someone here can try. The coding can be done within an hour.

**Comment by gcp** [comment]
> The coding can be done within an hour.

I look forward to someone testing your patch then. As already stated months ago, I won't make any changes to the training, but anyone can submit a new trained network for a test match (and promotion).

(I'm a bit at a loss as to where what you propose matches unrolling the network predictions as muZero does. I probably misunderstood either the muZero paper or what you're saying)

**Comment by Grant-Tao** [comment]
The above suggestion is just the simplest one based on current leela zero's training data and the training code.
A closer approach to MuZero would be minimizing the averaged losses of K consequential positions. that is, computing K batches of losses, average them, then use this averaged loss to compute the gradient and update the network. 

In order to avoid the unnecessary computing, the regularization item should be computed only once every K batches at the time when computing the averaged loss.

**Comment by gcp** [comment]
The big problem with that is that it mixes interdependent positions during the gradient update, which we already know gives way worse results for the value head training.

The comparison to MuZero is very misleading: when doing this position unrolling, MuZero is training both the policy/value network but also a state prediction network. Because it needs to train the latter, it has no choice but to consider immediately succeeding positions. This isn't something they do intentionally because it helps training, they have to do it because otherwise they can't train at all!

For the same reason, I'd say that

> Muzero's paper has shown the advantage of this training approach.

is rather misleading, because the results they get are from stacking two full networks on top of each other.

**Comment by Grant-Tao** [comment]
MuZero has three functions: presentation  h, dynamics g and prediction f.  h and g are 16 blocks resnets, f is a 20 blocks resnet. except for the very beginning, during MCTS, g and f will be used, they together form a 36 blocks resnet, almost the same size of leela zero network (40 blocks).

The loss function of MuZero has four parts: reward loss of g, value loss of f and policy loss of f, plus a regularization item. But for games like Go, the reward loss of dynamics 
 g is 0, thus g degrades to a state transfer function: given previous state s and action a, g gives the next hidden state.
 
Overall speaking, for Game Go, MuZero trains three networks from samples with S0 ... St previous states and At+1 ... At+k future actions, but the loss only related to the prediction function f (value and policy network). 

If Leela zero trains its network with K consequential positions (my second suggestion, averaging losses of K consequential position batches) , compared with MuZero, though the overall network structures are much different,  the loss functions are the same.

Maybe the training will show something interesting.

**Comment by gcp** [comment]
>MuZero has three functions: presentation h, dynamics g and prediction f. h and g are 16 blocks resnets, f is a 20 blocks resnet. except for the very beginning, during MCTS, g and f will be used, they together form a 36 blocks resnet, almost the same size of leela zero network (40 blocks).

That seems contrary to #2542, which claims f is just the network head.

**Comment by Grant-Tao** [comment]
@gcp You are right, I misundenstood the author's words "uses the same architecture as AlphaZero".  here is a simple MuZero implementation. https://github.com/YuriCat/MuZeroJupyterExample  @YuriCat

But I still doubted it. If @GCP is right, that means MuZero's 16 block network (ELO more than 5000) can beat Leela Zero's 40 block network.

**Comment by JanFre** [comment]
Do you know why do they call MCTS node statistics min_max_stats? Does it mean that during the selection not an avarage (V+R) indicator is used but rather some mini-max logic instead?

**Comment by Google9876** [comment]
MuZero is based on AlphaZero's powerful search and search-based strategy iterative algorithms, but integrates a learned model into the training step. MuZero also extends AlphaZero to a broader set of environments, including a single agent domain and non-zero rewards at intermediate time steps. The main idea of ​​the algorithm is to predict future behaviors that are directly related to planning (see Figure 1 below). The model takes the received observations (such as a Go board image or an Atari game screenshot) as input, and then converts it to a hidden state. Next, the hidden state is iteratively updated through a cyclic process that receives the previous hidden state and the assumed next operation. At each step, the model predicts strategies (such as actions to play), value functions (such as predicted winners), and instant rewards. The sole goal of end-to-end training of the model is to accurately estimate these three important quantities to match the improved policy estimates and the values ​​generated through searching and observed rewards. For the hidden state, there are no direct constraints and requirements to capture the information needed to reconstruct the original observations, which greatly reduces the amount of information for model maintenance and prediction; nor does it require the hidden state to match the unknown and true state in the environment; nor is it targeted at Other constraints on state semantics.

**Comment by Google9876** [comment]
![54c8ffd130124457b8b53e33f9f444c3](https://user-images.githubusercontent.com/60776843/74093924-b583d880-4b13-11ea-860b-d9d6e851117d.jpg)    Above picture

**Comment by Google9876** [comment]
In contrast, hidden states can represent states in any way related to predicting current and future values ​​and strategies. Intuitively, agents can create rules and dynamics internally for the most accurate planning.

**Comment by alreadydone** [comment]
Unfortunately, MuZero doesn't share information across branches of the tree (if it determines a ladder doesn't work or that a group is dead in one variation, same computation effort would be required to do that in every variation; MCTSNet may be able to do that due to learnable backup rule but it's not incorporated). Moreover, actions are still in one-one correspondence to (possibly illegal) moves; there are no "big actions" like playing out a ladder, capturing race, or attacking a group. I think Facebook's work on hierarchical decision making (by generating and following (natural) language instructions) has the potential to become the prototype of next-generation Go AI: to learn to build up and maintain a library of programs by itself, including its own search algorithm to replace MCTS (that allows such information sharing and big actions) and time management algorithm. Giving names to programs ("tools") would demand the ability of classification and language invention (maybe VQ-VAE would be appropriate), and extracting/distilling knowledge from the model itself to a program library / external toolbox demands self-awareness (or perhaps what Bengio calls consciousness), hopefully paving the way to AGI.

**Comment by Google9876** [comment]
![54c8ffd130124457b8b53e33f9f444c3](https://user-images.githubusercontent.com/60776843/74154966-a5bcdf00-4c4e-11ea-9252-e655dbce0bd4.jpg)
Figure 1: Planning, action and training with a trained model. (A) The way MuZero uses his model for planning; (B) The way MuZero works in the environment; (C) The way MuZero trains its model. Detailed MuZero algorithm At each time step t, using the previous observations O_1,…, O_t and future behaviors a_t + 1,…, a_t + k as conditions, through a model with parameter θ µ_θ, for each k = 1… K steps Make predictions. The model predicts three future quantities: strategy

**Comment by Google9876** [comment]
![4bbdb5aca47e43998ed06284db7e17fb](https://user-images.githubusercontent.com/60776843/74155009-b9684580-4c4e-11ea-9805-8c3646bbb641.jpg) 
Above picture

**Comment by Google9876** [comment]
Value function
![640?wx_fmt=jpeg tp=webp wxfrom=5 wx_lazy=1 wx_co=1](https://user-images.githubusercontent.com/60776843/74155370-722e8480-4c4f-11ea-82ab-36befce36d28.jpg)

**Comment by Google9876** [comment]
And instant rewards
![0aa779ac34f84f70b1fc0f1c1f754893](https://user-images.githubusercontent.com/60776843/74155441-925e4380-4c4f-11ea-822f-9068209c31fb.jpg)

**Comment by Google9876** [comment]
Where u. Represents the observed correct reward, π represents the strategy used to select real-time actions, and γ represents the discount function of the environment. At each time step t, the MuZero model is jointly characterized by a representation function, a dynamic function, and a prediction function. In this paper, researchers have made a precise characterization of dynamic functions. Strategy and value functions
![a5e2226aaf564f8f88c60f78d07dcf91](https://user-images.githubusercontent.com/60776843/74155570-c5083c00-4c4f-11ea-9a75-c8db0f009b20.jpg)

**Comment by Google9876** [comment]
Where u. Represents the observed correct reward, π represents the strategy used to select real-time actions, and γ represents the discount function of the environment. At each time step t, the MuZero model is jointly characterized by a representation function, a dynamic function, and a prediction function. In this paper, researchers have made a precise characterization of dynamic functions. Strategy and value functions
![a5e2226aaf564f8f88c60f78d07dcf91](https://user-images.githubusercontent.com/60776843/74155808-1fa19800-4c50-11ea-9cbb-993a68be226e.jpg)

**Comment by Google9876** [comment]
It is calculated based on the internal state s ^ k, which is similar to AlphaZero's joint strategy and value network. Given such a model, it is possible to find hypothetical future trajectories a ^ 1,…, a ^ k based on past observations O_1,…, O_t. For example, a simple search can easily select k-sequences of action that maximize the value function. More generally, we may be able to apply any MDP (Markov Decision Process) planning algorithm to internal rewards and state space derived from dynamic functions. For each hypothesized time step k, all parameters of the model are jointly trained, so that after k actual time steps, the strategy, value, and reward are accurately matched with their respective target values. Similar to AlphaZero, the promoted strategic targets are generated by Monte Carlo Tree (MCTS) search. The first goal is to minimize the error between the prediction strategy p ^ k_t and the search strategy π_t + k; the second goal is to minimize the error between the predicted value v ^ k_t and the value goal z_t + k; the third goal Is to minimize the error between the predicted reward r ^ k_t and the observed reward u_t + k. Finally, the L2 regularization term is added, resulting in the following total loss:
![865db1c3b71942919eff5da946a2cc48](https://user-images.githubusercontent.com/60776843/74155989-6a231480-4c50-11ea-842a-b26d0ba67dd0.jpg)

**Comment by Google9876** [comment]
Experimental results In the experiment, the researchers applied MuZero algorithm to classic board games such as Go, Chess, and Japanese Shogi, as a benchmark to challenge planning problems, and also applied to 57 games in the Atari game environment as a visually complex enhancement Benchmarks in the field of study. Figure 2 below shows the performance of the MuZero algorithm in each game training. In the game of Go, although the calculation amount of each node in the search tree is less than AlphaZero, the performance of MuZero still slightly exceeds AlphaZero. This suggests that MuZero may cache its own calculations in the search tree and leverage each additional application of the dynamic model to develop a deeper understanding of location.
![959a0c1df6fd493bac686a0b373e4626](https://user-images.githubusercontent.com/60776843/74156159-c128e980-4c50-11ea-9ee5-2f0f9b50f310.jpg)

**Comment by Google9876** [comment]
Figure 2: MuZero algorithm evaluation results in chess, Japanese shogi, Go, and Atari game training respectively. In chess, shogi, and go games, the abscissa indicates the number of training steps, and the ordinate indicates the Elo score.
![48bb8a49ea7a4b2ab2d26bdd61064787](https://user-images.githubusercontent.com/60776843/74156248-ec133d80-4c50-11ea-803a-aed1caff61e4.jpg)

**Comment by Google9876** [comment]
Table 1: MuZero in the Atari game compared to previous agents. The researchers showed the comparison results between MuZero and other agents under the large-scale (upper part of the table) and small-scale (lower part of the table) data settings, showing that MuZero has average scores, median scores, Env. Frames, training time, and Five evaluation indicators (red boxes) in the training step achieved new SOTA results. To understand the role of the model in MuZero, the researchers also focused on the following experiments in Go and Pac-Man Atari games. They first tested the scalability of planning on a typical planning problem for Go (Figure 3A below). In addition, they studied the scalability of planning in all Atari games (Figure 3B below). They then compared their model-based learning algorithm with other similar model-free learning algorithms (Figure 3C below).
![7e8fd67ab76e4ff4949b6644eed894e9](https://user-images.githubusercontent.com/60776843/74156378-2977cb00-4c51-11ea-859c-1a1cf173be41.jpg)
Figure 3: MuZero's evaluation results on Go, 57 Atari games, Pac-Man games.

**Comment by Google9876** [comment]
The MuZero model was evaluated in chess, Japanese shogi and Go, and it was found that the model can rival AlphaZero's performance beyond humans. Moreover, in this experiment, its predecessor AlphaZero learned the rules in advance.
